system/src/Grav/Framework/Cache/Adapter/FileStorage.php

```diff
@@ -0,0 +1,179 @@
+     * Path to Storage-directory.
+     *
+     * @var string
+     */
+    protected $directory;
+
+    /**
+     * Temporary filename.
+     *
+     * @var string
+     */
+    protected $tmp;
+
+    /**
+     * Initiate Storage
+     *
+     * @param string $directory       Path to directory.
+     * @param string $namespace       Name of something.
+     * @param int    $defaultLifetime Duration.
+     */
+    public function __construct($directory, $namespace = '', $defaultLifetime = null)
+    {
+        parent::__construct($namespace, $defaultLifetime ?: 31557600);
+        $this->directory = $directory;
+    }
+
+    /**
+     * Fetches a value from storage.
+     *
+     * @param string $key  The unique key of this item in storage.
+     * @param mixed  $miss Value to return if the key does not exist.
+     *
+     * @return mixed The value of the item from storage, or $miss if non-existant.
+     * 
+     * @throws \Psr\SimpleCache\InvalidArgumentException
+     */
+    public function doGet($key, $miss = false)
+    {
+        parent::validateKey($key);
+        $now = time();
+        $file = $this->directory . DIRECTORY_SEPARATOR . $key;
+
+        if (!file_exists($file) || !$h = @fopen($file, 'rb')) {
+            return $miss;
+        }
+
+        if ($now >= (int) $expiresAt = fgets($h)) {
+            fclose($h);
+            @unlink($file);
+        } else {
+            $i = rawurldecode(rtrim(fgets($h)));
+            $value = stream_get_contents($h);
+            fclose($h);
+            if ($i === $key) {
+                return unserialize($value);
+            }
+        }
+
+        return $miss;
+    }
+
+    /**
+     * Persists data in storage.
+     *
+     * @param string                 $key   The key of the item to store.
+     * @param mixed                  $value The value of the item to store.
+     * @param null|int|\DateInterval $ttl   The Time To Live value of this item.
+     *
+     * @return bool True on success and false on failure.
+     *
+     * @throws \Psr\SimpleCache\InvalidArgumentException
+     * @throws \Symfony\Component\Filesystem\Exception\IOExceptionInterface
+     */
+    public function doSet($key, $value, $ttl = 31556926)
+    {
+        parent::validateKey($key);
+        $expiresAt = time() + (int) $ttl;
+        $file = $this->directory . DIRECTORY_SEPARATOR . $key;
+
+        if (!is_writable($this->directory)) {
+            mkdir($this->directory, 0755, true);
+        } else {
+            try {
+                if ($this->tmp === null) {
+                    $this->tmp = $this->directory . DIRECTORY_SEPARATOR . uniqid('', true);
+                }
+
+                file_put_contents($this->tmp, $value, LOCK_EX);
+
+                if ($expiresAt !== null) {
+                    touch($this->tmp, $expiresAt);
+                }
+
+                return rename($this->tmp, $file);
+            } catch (\CacheException $e) {
+                throw new \CacheException($e);
```

Comment:
This won't work. use `throw $e;` instead.

Reply:
Changed to just `throw $e;`.